SocketRocket/SRWebSocket.m

```diff
@@ -563,14 +583,27 @@ - (void)_failWithError:(NSError *)error;
     });
 }
 
-- (void)_writeData:(NSData *)data;
+- (void)_writeData:(NSData *)data
 {
-    [self assertOnWorkQueue];
+    [self _writeData:data complete:NULL];
+}
 
+- (void)_writeData:(NSData *)data complete:(SRSendCompleteBlock)complete {
+    [self assertOnWorkQueue];
+    
     if (_closeWhenFinishedWriting) {
+        if (complete) {
+            complete(SRErrorWithCodeDescription(2134, @"socket is closed"));
+        }
         return;
     }
-
+    
+    SRDataCallback *record = [[SRDataCallback alloc] init];
+    record.completeBlock = complete;
+    NSUInteger location = dispatch_data_get_size(_outputBuffer);
+    record.range = NSMakeRange(location, [data length]);
+    _sendCallbacks[@([data hash])] = record;
```

Comment:
Looks like the most important part of this code review:

`hash` property on `NSObject`, and as a result `NSData`  is not guaranteed to be unique, meaning that in an unlikely event - you can run into collisions.

You need to change the key here to probably `[NSValue valueWIthRange:range]` and then to retrieve it - use the `NSValue` constructor once again.